std/net/curl.d

```diff
@@ -237,17 +237,52 @@ version(unittest)
         return initOnce!server(startServer());
     }
 
-    private immutable(T)[] recvAll(T=char)(Socket s)
+    private struct Request(T)
     {
-        ubyte[1024] buf=void;
-        ubyte[] res;
+        string hdrs;
+        immutable(T)[] bdy;
+    }
+
+    private Request!T recvReq(T=char)(Socket s)
+    {
+        ubyte[1024] tmp=void;
+        ubyte[] buf;
+
         while (true)
         {
-            auto nbytes = s.receive(buf[]);
+            auto nbytes = s.receive(tmp[]);
             assert(nbytes >= 0);
-            res ~= buf[0 .. nbytes];
-            if (nbytes < buf.length)
-                return cast(typeof(return))res;
+
+            immutable beg = buf.length > 3 ? buf.length - 3 : 0;
+            buf ~= tmp[0 .. nbytes];
+            auto bdy = buf[beg .. $].find(cast(ubyte[])"\r\n\r\n");
+            if (bdy.empty)
+                continue;
+
+            auto hdrs = cast(string)buf[0 .. $ - bdy.length];
+            bdy.popFrontN(4);
+            // no support for chunked transfer-encoding
+            if (auto m = hdrs.matchFirst(ctRegex!(`Content-Length: ([0-9]+)`, "i")))
+            {
+                import std.uni : asUpperCase;
+                if (hdrs.asUpperCase.canFind("EXPECT: 100-CONTINUE"))
+                    s.send(httpContinue);
+
+                size_t remain = m.captures[1].to!size_t - bdy.length;
```

Comment:
Maybe use the parsed content length to reduce the number of allocations for `buf`?